src/python/pants/core/util_rules/gen_rule.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+
+from pants.core.target_types import (
+    FilesSources,
+    GenRuleCommandField,
+    GenRuleOutputsField,
+    GenRuleSources,
+    GenRuleToolsField,
+)
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.fs import (
+    EMPTY_SNAPSHOT,
+    AddPrefix,
+    CreateDigest,
+    Digest,
+    Directory,
+    GlobMatchErrorBehavior,
+    MergeDigests,
+    PathGlobs,
+    Snapshot,
+)
+from pants.engine.process import (
+    BashBinary,
+    BinaryNotFoundError,
+    BinaryPathRequest,
+    BinaryPaths,
+    Process,
+    ProcessResult,
+    SearchPath,
+)
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    GeneratedSources,
+    GenerateSourcesRequest,
+    Sources,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+
+
+class GenerateFilesFromGenRuleRequest(GenerateSourcesRequest):
+    input = GenRuleSources
+    output = FilesSources
```

Comment:
Note that `Files` are not (currently) included when building a `pex_binary` and `python_awslambda`. They're mostly helpful for tests and for `archive`. Is that okay?

Reply:
Ah, I'd like the output to be as generic as possible, is plains `Sources` better, then? 
Or rather, I want to make sure that whatever is produced by the `shell_command` (if we call it that) may be used as sources by another target.

Not sure if simply depending on a `shell_command`, from say a `python_library` is enough, or if you should include the output files from `shell_command` in the `python_library` sources field? But that may complain that the "glob doesn't match" any files, until the `shell_command` has been executed, if that is an issue?